clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ArtifactController.java

```diff
@@ -66,4 +63,33 @@ StreamingResponseBody fetch(@RequestBody Artifact artifact) {
 
     return outputStream -> IOUtils.copy(artifactDownloader.download(artifact), outputStream);
   }
+
+  @RequestMapping(method = RequestMethod.GET, value = "/{type}/account/{accountName}/names")
+  List<String> getNames(@PathVariable("type") String type,
+                        @PathVariable("accountName") String accountName) {
+    ArtifactCredentials credentials = findArtifactCredentials(accountName);
+    if (!credentials.handlesType(type.replace("-","/"))) {
+      throw new IllegalArgumentException("Artifact credentials '" + accountName + "' cannot handle artifacts of type '" + type + "'");
+    }
+    return credentials.getArtifactNames();
+  }
+
+  @RequestMapping(method = RequestMethod.GET, value = "/{type}/account/{accountName}/names/{artifactName}/versions")
+  List<String> getVersions(@PathVariable("type") String type,
+                           @PathVariable("accountName") String accountName,
+                           @PathVariable("artifactName") String artifactName) {
+    ArtifactCredentials credentials = findArtifactCredentials(accountName);
+    if (!credentials.handlesType(type.replace("-","/"))) {
```

Comment:
This makes me a little nervous -- it's effectively doing the same thing that a url-encoding does, but instead of relying on a shared standard, it only applies to this endpoint.

What if we made `type` a query param? Also what's wrong with relying on url-encoding? Our http libraries are smart enough to handle the `/` on both ends (request & controller).

Reply:
@lwander, the server rejects requests and returns 400 HTTP error when it sees `%2F` in the request. This is default Tomcat behavior. `ALLOW_ENCODED_SLASH` boolean flag may be specified for Tomcat, but it starts to see `%2F` as delimiter. Anyway default implementation of HttpFirewall interface that Spring Security uses is the reason requests are rejected as well.

Reply:
Ah right -- is that only for path params or query params as well? Only asking in the hope that we can keep the API a little more readable/understandable

Reply:
Does `/artifacts/account/{accountName}/versions?<type>&<name>&<location>...` seem OK? it puts all identifying details aside from account into a query param, allowing a user to specify only the keys they feel are needed.

Reply:
@lwander Appreciate your suggestion, but I would like to clarify. Do you mean we should leave 2 API: 
`/artifacts/account/{accountName}/names?<type>&<location>&...` and
`/artifacts/account/{accountName}/versions?<type>&<name>&<location>...`

Or we should leave only one:
`/artifacts/account/{accountName}/versions?<type>&<name>&<location>...` and get all artifact names if `name` field is missed. If it's not - get artifact versions according to chosen name.

Nevertheless, I'm a little bit confused because `type` is still going after `artifactAccount`. But if we can't use `type` as path variable and there are no solutions, I think the best way is to leave API:
`/artifacts/account/{accountName}/names?<type>`
`/artifacts/account/{accountName}/versions?<type>`

Reply:
> Do you mean we should leave 2 API:
`/artifacts/account/{accountName}/names?<type>&<location>&...` and
`/artifacts/account/{accountName}/versions?<type>&<name>&<location>...`

That sounds good to me. 

> Nevertheless, I'm a little bit confused because type is still going after artifactAccount. 

I agree it seems a bit odd -- but in reality there is no reason why an artifactAccount can't download more than 1 type. Even in the helm case, if you have a repository in GCS or S3, you should (in the future) be able to use a GCS or S3 account to read the contents of the repository, knowing that it is a proper helm repo.

Reply:
@lwander, please take a look at changes I made according to last replies.